File: src/commands/define.ts
import {
    AttachmentBuilder, EmbedBuilder,
    SlashCommandBuilder
} from 'discord.js';
import { ErrorEmbed } from '../core/functions.js';
import type { Command } from '../typings/core.js';
import type { dictionaryAPI } from '../typings/definitions.js';

export default <Command>{
    data: new SlashCommandBuilder()
        .setName("define")
        .setDescription("Get the definition for a word")
        .addStringOption((Option) =>
            Option
                .setName("word")
                .setDescription("Type the word you want definition for")
                .setRequired(true)
        ),
    async execute(interaction) {
        await interaction.deferReply();

        const query = interaction.options.getString("word", true).trim();

        const response = await fetch("https://api.dictionaryapi.dev/api/v2/entries/en/" + query);
        if (!response.ok)
            throw ErrorEmbed("Dictionary API Error", `The server declined our request with reason: \`${response.statusText}\``);

        const data = await response.json() as dictionaryAPI[];

        if (!data[0] || !data[0].meanings[0] || !data[0].meanings[0].definitions[0])
            throw ErrorEmbed("Dictionary API Error", `Definition not found for the word: "${query}"`);

        const phoneticsText = data[0]!.phonetics.find(p => p.text)?.text ?? `No phonetics available`;
        const phoneticsAudio = data[0]!.phonetics.find(p => p.audio)?.audio;

        const word = data[0]!.word ?? query;

        let text = `# ${word.charAt(0).toUpperCase() + word.slice(1)}\n`;
        text += `${phoneticsText}\n\n\n`;

        for (const meaning of data[0]!.meanings) {
            text += `**${meaning.partOfSpeech}**\n`;
            text += meaning.definitions.map((def, index, arr) => {
                let temp = ``;

                if (arr.length > 1) {
                    temp += `### ${index + 1}. ${def.definition}\n`;
                    temp += def.example ? `-# "${def.example}"` : ``;
                }

                else temp += `### ${def.definition}\n${def.example ? `-# "${def.example}"` : ``}`;

                return temp;
            }).join('\n');

            text += `\n\n`;
        }

        let attachment: AttachmentBuilder | null = null;

        if (phoneticsAudio) {
            const response = await fetch(phoneticsAudio);
            attachment = new AttachmentBuilder(
                Buffer.from(await response.arrayBuffer()),
                { name: `${word}.mp3` }
            );
        }

        await interaction.editReply({
            embeds: [
                new EmbedBuilder()
                    .setDescription(text)
                    .setColor("Green")
            ],
            files: attachment ? [attachment] : []
        });
    },
}

File: Commands/define.py
import io
import requests
from disnake.ext.commands import slash_command, InteractionBot
from disnake import (
    ApplicationCommandInteraction,
    Embed,
    File,
    Option,
    OptionType,
)

@slash_command(
    name="define",
    description="Get the definition for a word",
    options=[
        Option(
            name="word",
            description="Type the word you want definition for",
            required=True,
            type=OptionType.string
        )
    ]
)
async def define_command(interaction: ApplicationCommandInteraction, word: str):
    await interaction.response.defer()

    query = word.strip()
    url = f"https://api.dictionaryapi.dev/api/v2/entries/en/{query}"

    try:
        response = requests.get(url)
    except Exception as e:
        return await interaction.edit_original_response(
            embed=Embed(
                title="Dictionary API Error",
                description=f"Failed to connect to the API.\n`{e}`",
                color=0xFF0000,
            )
        )

    if response.status_code != 200:
        return await interaction.edit_original_response(
            embed=Embed(
                title="Dictionary API Error",
                description=f"The server declined our request: `{response.reason}`",
                color=0xFF0000,
            )
        )

    data = response.json()

    if not isinstance(data, list) or not data or not data[0].get("meanings"):
        return await interaction.edit_original_response(
            embed=Embed(
                title="Dictionary API Error",
                description=f"Definition not found for the word: **{query}**",
                color=0xFF0000,
            )
        )

    entry = data[0]
    phonetics = entry.get("phonetics", [])
    phonetic_text = next((p.get("text") for p in phonetics if p.get("text")), "No phonetics available")
    phonetic_audio = next((p.get("audio") for p in phonetics if p.get("audio")), None)
    word_display = entry.get("word", query)

    # --- Build description ---
    text = f"# {word_display.capitalize()}\n{phonetic_text}\n\n"

    for meaning in entry.get("meanings", []):
        part_of_speech = meaning.get("partOfSpeech", "unknown")
        text += f"**{part_of_speech}**\n"
        definitions = meaning.get("definitions", [])
        for i, definition in enumerate(definitions):
            def_text = definition.get("definition", "No definition found.")
            example = definition.get("example")
            if len(definitions) > 1:
                text += f"### {i + 1}. {def_text}\n"
            else:
                text += f"### {def_text}\n"
            if example:
                text += f"-# \"{example}\"\n"
        text += "\n"

    embed = Embed(description=text, color=0x00FF00)

    # --- Handle audio attachment ---
    if phonetic_audio:
        try:
            audio_response = requests.get(phonetic_audio)
            audio_bytes = io.BytesIO(audio_response.content)
            file = File(audio_bytes, filename=f"{word_display}.mp3")
            await interaction.edit_original_response(embed=embed, file=file)
            return None
        except Exception as e:
            print(f"Audio fetch error: {e}")

    await interaction.edit_original_response(embed=embed)
    return None


def setup(bot: InteractionBot):
    bot.add_slash_command(define_command)
